### ISSUE TITLE
Custom dataloader yolo3/ssd/frcnn, training loss decreases but validation map remains 0 

### ISSUE THREAD

**Issue by djaym7** [dmlc/gluon-cv#1512]
```
import imgaug as ia
import imgaug.augmenters as iaa
import numpy as np
import mxnet as mx
import gluoncv
import copy
from gluoncv.data.transforms import bbox as tbbox
from gluoncv.data.transforms import image as timage
from gluoncv.data.transforms import experimental


sometimes = lambda aug: iaa.Sometimes(0.5, aug)

seq = iaa.Sequential(
    [
        # apply the following augmenters to most images
        iaa.Fliplr(0.5), # horizontally flip 50% of all images
        iaa.Flipud(0.2), # vertically flip 20% of all images
        # crop images by -5% to 10% of their height/width
        sometimes(iaa.CropAndPad(
            percent=(-0.05, 0.1),
            pad_mode=ia.ALL,
            pad_cval=(0, 255)
        )),
        sometimes(iaa.Affine(
            scale={"x": (0.3, 1.5), "y": (0.3, 1.5)}, # scale images to 80-120% of their size, individually per axis
            translate_percent={"x": (-0.2, 0.2), "y": (-0.2, 0.2)}, # translate by -20 to +20 percent (per axis)
            rotate=(-10, 10), # rotate by -45 to +45 degrees
            shear=(-5, 5), # shear by -16 to +16 degrees
            order=[0, 1], # use nearest neighbour or bilinear interpolation (fast)
            cval=(0, 255), # if mode is constant, use a cval between 0 and 255
            mode='constant' # use any of scikit-image's warping modes (see 2nd image from the top for examples)
        ))
    ],
    random_order=True
)


def aug(img,labels):
    if labels.shape[1]>5:
        bboxes=[ia.BoundingBox(xmax,ymax,xmin,ymin,label) for xmin,ymin,xmax,ymax,label in labels[:,:5]]
        images_aug, bbs_aug = seq(images=img.expand_dims(0).asnumpy(), bounding_boxes=bboxes)
        bbs_aug = np.array([[i.x2,i.y2,i.x1,i.y1,i.label,_] for i,_ in zip(bbs_aug,labels[:,5:])])
    else:
        bboxes=[ia.BoundingBox(xmax,ymax,xmin,ymin,label) for xmin,ymin,xmax,ymax,label in labels]
        images_aug, bbs_aug = seq(images=img.expand_dims(0).asnumpy(), bounding_boxes=bboxes)
        bbs_aug = np.array([[i.x2,i.y2,i.x1,i.y1,i.label] for i in bbs_aug])
    return images_aug[0],bbs_aug







class YOLO3DefaultTrainTransform(object):
    """Default YOLO training transform which includes tons of image augmentations.
    Parameters
    ----------
    width : int
        Image width.
    height : int
        Image height.
    net : mxnet.gluon.HybridBlock, optional
        The yolo network.
        .. hint::
            If net is ``None``, the transformation will not generate training targets.
            Otherwise it will generate training targets to accelerate the training phase
            since we push some workload to CPU workers instead of GPUs.
    mean : array-like of size 3
        Mean pixel values to be subtracted from image tensor. Default is [0.485, 0.456, 0.406].
    std : array-like of size 3
        Standard deviation to be divided from image. Default is [0.229, 0.224, 0.225].
    iou_thresh : float
        IOU overlap threshold for maximum matching, default is 0.5.
    box_norm : array-like of size 4, default is (0.1, 0.1, 0.2, 0.2)
        Std value to be divided from encoded values.
    """
    def __init__(self, width, height, net=None, mean=(0.485, 0.456, 0.406),
                 std=(0.229, 0.224, 0.225), mixup=False, **kwargs):
        self._width = width
        self._height = height
        self._mean = mean
        self._std = std
        self._mixup = mixup
        self._target_generator = None
        if net is None:
            return

        # in case network has reset_ctx to gpu
        self._fake_x = mx.nd.zeros((1, 3, height, width))
        net = copy.deepcopy(net)
        net.collect_params().reset_ctx(None)
        with mx.autograd.train_mode():
            _, self._anchors, self._offsets, self._feat_maps, _, _, _, _ = net(self._fake_x)
        from gluoncv.model_zoo.yolo.yolo_target import YOLOV3PrefetchTargetGenerator
        self._target_generator = YOLOV3PrefetchTargetGenerator(
            num_class=len(net.classes), **kwargs)

    def __call__(self, src, label):
        """Apply transform to training image/label."""
#######################################################################################
        img,bbox=aug(src,label)
        img = mx.nd.array(img,dtype=np.float32)

        # resize with random interpolation
        h, w, _ = img.shape
        interp = np.random.randint(0, 5)
        img = timage.imresize(img, self._width, self._height, interp=interp)
        bbox = tbbox.resize(bbox, (w, h), (self._width, self._height))
        
        # to tensor
        img = mx.nd.image.to_tensor(img)
        img = mx.nd.image.normalize(img, mean=self._mean, std=self._std)
#         print(type(img),img.dtype,type(bbox),bbox.dtype,img.max(),img.min())
# ################################################################################
        if self._target_generator is None:
            return img, bbox
        # generate training target so cpu workers can help reduce the workload on gpu
        gt_bboxes = mx.nd.array(bbox[np.newaxis, :, :4])
        gt_ids = mx.nd.array(bbox[np.newaxis, :, 4:5])
        if self._mixup:
            gt_mixratio = mx.nd.array(bbox[np.newaxis, :, -1:])
        else:
            gt_mixratio = None
        objectness, center_targets, scale_targets, weights, class_targets = self._target_generator(
            self._fake_x, self._feat_maps, self._anchors, self._offsets,
            gt_bboxes, gt_ids, gt_mixratio)
        return (img, objectness[0], center_targets[0], scale_targets[0], weights[0],
                class_targets[0], gt_bboxes[0])

    
    
```   

The training errors are getting low but the validation is always 0. Validation uses the default transform in the script and everything else is same too as train yolo3 py

**Comment by zhreshold** [comment]
applying iaa.Flipud(0.2) is questionable, other than that it look fine to me, will you be able to verify the  corrrectness of `img,bbox=aug(src,label)`, by visulization or something similar?

**Comment by djaym7** [comment]
Testing aug works fine, test is shown below
'''
#when  target_generator is None in transform
from gluoncv.data.lst.detection import LstDetection
data = LstDetection('../Segmentation Job2/LST/out.lst',root='../Segmentation Job2/Unlabelled')
image,labels=data[0]
print(image.shape,labels[0])
image,labels = transform(image,labels)
image = image.transpose((1,2,0))

print(image.shape,labels[0])
# labels[:,0:2],labels[:,2:4] = labels[:,2:4],labels[:,0:2]
plt.imshow(cv_plot_bbox(image,bboxes=labels[:,:-1],labels=labels[:,-1],scale=1,linewidth=1))
'''

**Comment by djaym7** [comment]
![image](https://user-images.githubusercontent.com/12378820/98886094-2dd34d80-2448-11eb-9697-0db511eb6887.png)

The outputs of both the transforms (original and custom) have same dtypes, max and mins and plotting them also has no problems. Using just horizontal flips in the custom transform still gives the same problem : training loss decreases but val mAP=0 for all the objects VS in original transform its converging for both train and val

**Comment by djaym7** [comment]
Solved and deleted. Thanks
